docs/notebooks/mixed_search_spaces.pct.py

```diff
@@ -0,0 +1,207 @@
+
+# %%
+from trieste.space import Box, DiscreteSearchSpace
+
+minimizers0 = ScaledBranin.minimizers[:, 0]
+step = (minimizers0[1] - minimizers0[0]) / 4
+points = np.concatenate(
+    [
+        # Equally spaced points to the left of the middle minimizer. Skip the last point as it is
+        # the same as the first point in the next array.
+        np.flip(np.arange(minimizers0[1], 0.0, -step))[:-1],
+        # Equally spaced points to the right of the middle minimizer.
+        np.arange(minimizers0[1], 1.0, step),
+    ]
+)
+discrete_space = DiscreteSearchSpace(points[:, None])
+continuous_space = Box([0.0], [1.0])
+mixed_search_space = discrete_space * continuous_space
+
+
+# %% [markdown]
+# Next we demonstrate the Branin function over the mixed search space, by plotting the original
+# function contours and highlighting the discretization points.
+# The discrete dimension is along the x-axis and the continuous dimension is on the y-axis, with
+# the vertical dashed lines indicating the discretization points.
+
+# %%
+from trieste.experimental.plotting import plot_function_2d
+
+# Plot over the predefined search space.
+fig, ax = plot_function_2d(
+    scaled_branin,
+    ScaledBranin.search_space.lower,
+    ScaledBranin.search_space.upper,
+    contour=True,
+)
+
+ax[0, 0].set_xlabel(r"$x_1$")
+ax[0, 0].set_ylabel(r"$x_2$")
+
+# Draw vertical lines at the discrete points.
+for point in points:
+    ax[0, 0].vlines(
+        point,
+        mixed_search_space.lower[1],
+        mixed_search_space.upper[1],
+        colors="b",
+        linestyles="dashed",
+        alpha=0.6,
+    )
+
+# %% [markdown]
+# ## Sample the observer over the search space
+#
+# We begin our optimization by first collecting five function evaluations from random locations in
+# the mixed search space.
+
+# %%
+from trieste.objectives import mk_observer
+
+observer = mk_observer(scaled_branin)
+
+num_initial_points = 5
+initial_query_points = mixed_search_space.sample(num_initial_points)
+initial_data = observer(initial_query_points)
+
+# %% [markdown]
+# ## Model the objective function
+#
+# We now build a Gaussian process model of the objective function using the initial data, similar
+# to the [introduction notebook](expected_improvement.ipynb).
+
+# %%
+from trieste.models.gpflow import GaussianProcessRegression, build_gpr
+
+gpflow_model = build_gpr(
+    initial_data, mixed_search_space, likelihood_variance=1e-7
+)
+model = GaussianProcessRegression(gpflow_model)
+
+# %% [markdown]
+# ## Run the optimization loop
+#
+# The Bayesian optimization loop is run for 15 steps over the mixed search space.
+# For each step, the optimizer fixes the discrete dimension to the best points found from a random
+# initial search, and then optimizes the continuous dimension using a gradient-based method.
+
+# %%
+from trieste.bayesian_optimizer import BayesianOptimizer
+
+bo = BayesianOptimizer(observer, mixed_search_space)
+
+num_steps = 15
```

Comment:
I think we add some bits in `quickrun` folder to reduce the build time for quick tests, that reduces e.g. num steps to 1 or 2, see other examples in that folder...